### DISCUSSION TITLE
Configure CSV-Export

### DISCUSSION THREAD

**Discussion by ooleanderoo** [kimai/kimai#2521] [Questions & Answers]
Hi,

I'm searching for the possibility to adjust the CSV-Export.
We are currently testing Kimai and now my aim is to have the possibility to export the Timesheets to .csv to be able to import it to our accounting-software DATEV.
There is not much to do I think. I already am able to adjust the importer of DATEV a little but there are some little things and one tricky one that has to be made already by kimai I think.

Here is an example of an importable file for DATEV
```
######	#	05/2021								
14	640	1	8501		2.00	1.00	18013			
14	640	1	8501		6.00		19013			
15	356	1	901		3.00	1.00	20007			
15	356	1	901		5.00		20008			
15	356	1	1500		1.50		20008			
15	356	1	1500		1.50		20007	
```	

Here is an example of an importable file for DATEV with comments
```
######//static company related number//	#//static company related number//	05/2021 //month and year of import-data ------- The First Line is a header and always needed//							

//day of month (1-31)// //number of employee//	//Key for ill-days//	//activity//	//duration//	//first entry of day gets a 1.00//	//project//
14			640			1			8501		2.00		1.00					18013			
14			640			1			8501		6.00							19013			
15			356			1			901		3.00		1.00					20007			
15			356			1			901		5.00							20008			
15			356			1			1500		1.50							20008			
15			356			1			1500		1.50							20007		
```	
										

The example files are tab-seperated. In DATEV you can change it also to comma or semicolon seperated data (and more... like the import-data function in Excel). And you also can tell DATEV which column belongs to which data.
The example files have the minimum data needed to import the timesheets.

I think the somehow tricky entry is that the first entry of a day has to get an extra entry with a value of 1.00 (like "Has worked"->yes)

The "Illness-Key" can be provided by a custom field of an activity or project i think. The number of the employee as well.

The header probably also needs some extra doing... But for the header i already thought about doing it manualy afterwards. Not the best way but this could be a future feature. We are currently in the state of testing....

So my question now is just how complex is that or is there possibly already a way to do it on my own? Haven't read about something like this in the documentation ;) 

Best Regards!

**Comment by kevinpapst** [comment]
You can add your own export renderer:
https://www.kimai.org/documentation/export.html#adding-export-renderer

Take the demo bundle as an example for your own bundle:
https://github.com/Keleo/DemoBundle

You can use the https://github.com/kevinpapst/kimai2/blob/master/src/Export/Base/AbstractSpreadsheetRenderer.php for your own CSV implementation - [start debugging here](https://github.com/kevinpapst/kimai2/blob/master/src/Export/Renderer/CsvRenderer.php).

I am available for freelance work as well.

**Reply by ooleanderoo** [reply]
Thanks for the fast reply. I will have a look at it on the weekend ;). Maybe I'll get something to work to be able to test it a little bit.

Propably you'll get feedback than ;)

**Comment by HeinzWuert** [comment]
Ich glaube das löst das eigentliche, verdeckte Problem nicht.
Grund ist, dass DB Auswertungen laufen müssen um die geforderten Variablen zu füllen. Das Abfragen einzelner Datensätze hilft nicht. Es werden Summenzüge benötigt die zuerst durch Kimai errechnet und dann in einer CSV Ausgabe zur Verfügung gestellt werden müssen.
Ausserdem fehlt in der DB-Struktur von Kimai z.B. noch das Datenfeld für die Rechnungsnr des Nutzers, welches ebenfalls an die FiBu  zu übertragen ist. Ohne müsste man nach dem Export alle Datensätze händisch um diesen Parameter ergänzen.
Die Schnittstelle zu einer Fibu bedarf daher sowohl einer Auswertungsroutine als auch Ergänzungen in der DB-Struktur von Kimai, aber wir arbeiten ja daran ;-)
Ob dann der Export für DATEV oder wie in unserem Fall Lexware erfolgt wird dann egal sein. Das Anfordserungsprofil ist das Gleiche. Sinnig wäre natürlich hier verschiedene Anwender ins Boot zu holen, um mit einer Schnittstelle alle versorgen zu können.
